src/Geography.cpp

```diff
@@ -0,0 +1,236 @@
+    for (RingConstIterator idx = m_rings.begin();
+         idx != m_rings.end();
+         idx++) {
+        // Each ring has 3 coordinates for each point
+        // plus 5 + 38 bytes of overhead.  Note that we
+        // will not serialize the duplicated first point.
+        answer += (idx->numPoints()-1)*3*sizeof(double) + 5 + 38;
+    }
+    return answer;
+}
+
+int32_t Geography::serializeTo(ByteBuffer &buffer) const {
+    // Remember where we start, so that we can
+    // write the size.  We keep track of how much
+    // we actually write and return it, so that the
+    // caller can check that this and getSerializedSize
+    // agree.
+    int32_t start = buffer.position();
+    if (isNull()) {
+        buffer.putInt32(-1);
+        return 4;
+    }
+
+    // Save space for the size.  We don't
+    // really know what it is now, so we
+    // well seek back to it after we are done.
+    buffer.putInt32(0);
+    // We write three bytes of overhead, followed by the
+    // number of rings.
+    buffer.putInt8(0);
+    buffer.putInt8(1);
+    buffer.putInt8(0);
+    buffer.putInt32(m_rings.size());
+    bool reverseit = false;
+    for (RingConstIterator idx = m_rings.begin();
+         idx != m_rings.end();
+         idx++) {
+        // Serialize each ring.
+        idx->serializeTo(buffer, reverseit);
+        reverseit = true;
+    }
+
+    // We write 33 bytes of overhead.
+    char zeros[33];
+    memset(zeros, 0, sizeof(zeros));
+    buffer.put(zeros, sizeof(zeros));
+
+    // Write the size where we started.
+    int32_t size = buffer.position() - start;
+    buffer.putInt32(start, size-4);
+    return size;
+}
+
+int32_t Geography::deserializeFrom(ByteBuffer &aData,
+                                   int32_t first_offset,
+                                   bool &aWasNull)
+{
+    int32_t offset = first_offset;
+    // The first four bytes are the length.  We read them
+    // only so we can validate the read.  The length does
+    // not include them.
+    int32_t len = aData.getInt32(offset);
+    if (len == -1) {
+        aWasNull = true;
+        makeNull();
+        return 4;
+    } else {
+        aWasNull = false;
+    }
+    int32_t last_offset = first_offset + 4 + len;
+    offset += sizeof(int32_t);
+    assert(offset < last_offset);
+    // The first seven bytes are somewhat overhead, but they contain
+    // the number of rings.
+    int32_t numRings = aData.getInt32(offset + 3);
+    offset += 7;
+    assert(offset < last_offset);
+    assert(0 <= numRings);
+    for (int idx = 0; idx < numRings; idx += 1) {
+        Ring &rng = addEmptyRing();
+        rng.clear();
+        // The first byte is noise.
+        offset += 1;
+        assert(offset < last_offset);
+        // The next 4 bytes are the number of vertices.
+        int numVerts = aData.getInt32(offset);
+        offset += 4;
+        assert(offset < last_offset);
+        assert(3 <= numVerts);
+        for (; numVerts > 0; numVerts -= 1) {
```

Comment:
No.  There are 3 coordinates per vertex, but the number of vertices of a rectangle, for example is 5.  There is an extra one which closes the loop.